### PR TITLE
ES 6.2.2

### PR DESCRIPTION
**Pull request by softwaredoug**
TODO

- [x] Figure out how to deal with `tf` in ExplorerScorer
- [x] Review whether isCacheable is correct for each weight we implement

### CODE REVIEW

File: src/main/java/com/o19s/es/ltr/query/RankerQuery.java [hunk 1]
```diff
@@ -186,6 +191,13 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
     public class RankerWeight extends Weight {
         private final List<Weight> weights;
 
+        @Override
+        public boolean isCacheable(LeafReaderContext ctx) {
+            // Only cachable if model doesn't change?
+            // I *think* it should be the case it should be true
+            return true;
```
Comment (nomoa) [inline src/main/java/com/o19s/es/ltr/query/RankerQuery.java]: I'd go with the BooleanWeight approach here: https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java#L314

**Comment by jettro** [comment]
It seems the freq() method of Scorer has been removed from the parent, but some scorers still have it. TermScorer and ExactPhraseScorer for instance. Not sure why yet. But here is the Lucene commit that makes the change:
https://github.com/apache/lucene-solr/commit/183571c085201a2f1e126852647910d1930d50a6

Also not su much a solution, more part of an explanation:
https://issues.apache.org/jira/browse/LUCENE-6278

Can we maybe use the elastic endpoint term vectors here?
https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-termvectors.html

File: src/main/java/com/o19s/es/ltr/query/RankerQuery.java (hunk 1 above)
Comment (nomoa) [reply src/main/java/com/o19s/es/ltr/query/RankerQuery.java]: So checked a bit more what's happening here:
- the valid way is to delegate to sub weights like what is done the BooleanWeight

But in fact it's just called to see if lucene can cache the docIds if this query was used as a filter.
So could very much just `throw new IllegalStateException("RankerWeight should not be used as a filter and cached");`

**Comment by nomoa** [comment]
For the freq() problem I'd go with the easy way and check if the child scorer is a TermScorer instance then cast so that you can access the freq() method there as it's still available (as suggested by jettro).

The hard way (we'll have to do it if they remove freq() from TermScorer) would be:
- instead of accepting an arbitrary query we'll accept a query string with field or an analyzer
- tokenize the query based on the analyzer provided
- Scan the freq() from the postings

Using termVector here is not option imo, it'd be too slow.

**Comment by softwaredoug** [comment]
@nomoa I think if we do that (w/ freq), we're officially supporting only the [ES term query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html) and we're assuming Elasticsearch will always map their term query to Lucene's term query?

**Comment by softwaredoug** [comment]
Looks like it may not be possible to check if scorer is a TermScorer, TermScorer appears to be [package private](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TermScorer.java#L29)

**Comment by nomoa** [comment]
arg... if TermScorer is package private I don't have easy solution to suggest... :(
Going the hard way is much more code to write...

**Comment by nomoa** [comment]
LGTM on my side, the build pass locally.
CI needs to be upgraded to java9.

**Comment by softwaredoug** [comment]
Upgraded Circle to 2.0 to get a Java 9 (a yak shave, within a yak shave :) ). Seems to work

**Comment by nomoa** [comment]
LGTM, I'd just update the PR to remove the WIP flag
The last failure seems to be related to a JVM bug, if possible I'd check if a hs_err_pidXXX.log was generated somewhere but not sure that CircleCI keeps an image of the FS somewhere of the failed builds.

**Comment by purbon** [comment]
test pass for me locally (mac os x)
screen shot 2018-03-20 at 15 20 41

can we retrigger the build in cicle CI and join the build with ssh to try to investigate the error a bit more?

**Comment by softwaredoug** [comment]
Retriggering didn't work, following up on the circle support forums. It appears to be related to a JVM not having enough memory.

**Comment by softwaredoug** [comment]
Build eventually passed...

**Pull request merged**